go_covert_lib/controller/channel/httpCovert/httpCovert.go

```diff
@@ -0,0 +1,288 @@
+package httpCovert
+
+import (
+	"context"
+	"errors"
+	"log"
+	"net"
+	"net/http"
+	"strconv"
+	"time"
+)
+
+const maxMsg = 32
+const (
+	Client = 0
+	Server = 1
+)
+
+// This is a normal, covert HTTP messaging channel
+// The message is sent using HTTP packets with the proper OS tcp functions
+type Config struct {
+	FriendIP   [4]byte
+	OriginIP   [4]byte
+	FriendPort uint16
+	OriginPort uint16
+	UserType   uint8
+
+	ClientPollRate time.Duration
+	ClientTimeout  time.Duration
+	ReadTimeout    time.Duration
+	WriteTimeout   time.Duration
+}
+
+// A HTTP covert channel
+type Channel struct {
+	conf Config
+	srv  *http.Server
+	clt  *http.Client
+
+	cancel chan bool
+
+	serverSendBuf chan []byte
+	serverRecBuf  chan []byte
+}
+
+// Create the covert channel, filling in the SeqEncoder
+// with a default if none is provided
+// Although the error is not yet used, it is anticipated
+// that this function may one day be used for validating
+// the data structure
+func MakeChannel(conf Config) (*Channel, error) {
+	c := &Channel{conf: conf, cancel: make(chan bool)}
+
+	// if the channel has been specified as the server
+	if c.conf.UserType == Server {
+		mux := http.NewServeMux()
+		mux.HandleFunc("/", c.handleFunc)
+
+		// launch a new http server
+		c.srv = &http.Server{Handler: mux}
+
+		// starts up the http server to listen
+		l, err := net.Listen("tcp4", ":"+strconv.Itoa(int(c.conf.OriginPort)))
+		if err != nil {
+			return nil, err
+		}
+
+		go func() { c.srv.Serve(l) }()
+
+		//buffers used to hold the server sent messages and received messages
+		c.serverSendBuf = make(chan []byte, maxMsg)
+		c.serverRecBuf = make(chan []byte, maxMsg)
+	}
+	return c, nil
+}
+
+//Handler function that is called everytime a new http request is received
+func (c *Channel) handleFunc(w http.ResponseWriter, r *http.Request) {
+
+	//if the request message is a get request
+	//extract the data from the servers send buffer and send it to the client
+	if r.Method == "GET" {
+		select {
+		case data := <-c.serverSendBuf:
+			w.Header().Add("Cookie", string(data))
```

Comment:
It is weird that "Valid" is set to zero if there is valid data present. It would normally be the other way around.

Reply:
resolved